eth/state_accessor.go

```diff
@@ -155,28 +178,28 @@ func (eth *Ethereum) StateAtBlock(block *types.Block, reexec uint64, base *state
 		nodes, imgs := database.TrieDB().Size()
 		log.Info("Historical state regenerated", "block", current.NumberU64(), "elapsed", time.Since(start), "nodes", nodes, "preimages", imgs)
 	}
-	return statedb, nil
+	return statedb, func() { database.TrieDB().Dereference(block.Root()) }, nil
```

Comment:
I'm unsure about this. You're returning a deref call, but there's no matching ref call added. Won't this end up us dereferencing something live?

Reply:
The reference work is done in the for loop.

Once we enter the loop for executing blocks, we will always reference the current one and reference the parent. And we can ensure *at least one block* is executed. In the end, we will keep the reference of the latest one.

Reply:
Ah ok, I didn't get yet to the part of TraceChain where you removed the dereference and replaced it with this :+1: